compiler/modulegraphs.nim

```diff
@@ -343,31 +364,44 @@ proc hash*(u: SigHash): Hash =
 
 proc hash*(x: FileIndex): Hash {.borrow.}
 
+template getC(): untyped =
+  when c is PContext: c
+  else: c.c
+
+template onDefAux(info: TLineInfo; s0: PSym, c0: untyped, isFwd: bool) =
+  if s0.kind in ExportableSymKinds:
+    let c = c0 # in case c0 is an expression
+    var top = true
+    case s0.kind
+    of skProc, skMacro:
+      # unfortunately, can't use `c.isTopLevel` because the scope isn't closed yet
+      top = c.currentScope.depthLevel <= 3
+    else: top = c.currentScope.depthLevel <= 2
+    if top and c.module != nil:
+      strTableAdd(semtabAll(c.graph, c.module), s0)
+      if c.config.symbolFiles != disabledSf:
+        addHidden(c.encoder, c.packedRepr, s0)
+
 when defined(nimfind):
   template onUse*(info: TLineInfo; s: PSym) =
-    when compiles(c.c.graph):
-      if c.c.graph.onUsage != nil: c.c.graph.onUsage(c.c.graph, s, info)
-    else:
-      if c.graph.onUsage != nil: c.graph.onUsage(c.graph, s, info)
+    let c2 = getC()
+    if c2.graph.onUsage != nil: c2.graph.onUsage(c2.graph, s, info)
 
   template onDef*(info: TLineInfo; s: PSym) =
-    when compiles(c.c.graph):
-      if c.c.graph.onDefinition != nil: c.c.graph.onDefinition(c.c.graph, s, info)
-    else:
-      if c.graph.onDefinition != nil: c.graph.onDefinition(c.graph, s, info)
+    let c2 = getC()
+    onDefAux(info, s, c2, false)
+    if c2.graph.onDefinition != nil: c2.graph.onDefinition(c2.graph, s, info)
 
   template onDefResolveForward*(info: TLineInfo; s: PSym) =
-    when compiles(c.c.graph):
-      if c.c.graph.onDefinitionResolveForward != nil:
-        c.c.graph.onDefinitionResolveForward(c.c.graph, s, info)
-    else:
-      if c.graph.onDefinitionResolveForward != nil:
-        c.graph.onDefinitionResolveForward(c.graph, s, info)
+    let c2 = getC()
+    onDefAux(info, s, c2, true)
+    if c2.graph.onDefinitionResolveForward != nil:
+      c2.graph.onDefinitionResolveForward(c2.graph, s, info)
 
 else:
   template onUse*(info: TLineInfo; s: PSym) = discard
-  template onDef*(info: TLineInfo; s: PSym) = discard
-  template onDefResolveForward*(info: TLineInfo; s: PSym) = discard
+  template onDef*(info: TLineInfo; s: PSym) = onDefAux(info, s, getC(), false)
+  template onDefResolveForward*(info: TLineInfo; s: PSym) = onDefAux(info, s, getC(), true)
```

Comment:
Instead of re-using the old, deprecated ``onDef`` mechanism, can't we patch `lookups.addInterfaceDeclAux`?

Reply:
done; it required a small refactoring in semEnum to call addInterfaceDeclAux instead of exportSym